src/main.rs

```diff
@@ -237,315 +241,277 @@ impl Format {
-                ),
-            },
-            Format::Repeat(a) => Format::Alt(
-                Box::new(Format::Empty),
-                Box::new(Format::Cat(a.clone(), Box::new(Format::Repeat(a.clone())))),
-            )
-            .might_match_lookahead(input, next),
-            Format::RepeatCount(_expr, _a) => {
-                true // FIXME
-            }
-            Format::Slice(_expr, _a) => {
-                true // FIXME
-            }
-            Format::WithRelativeOffset(_, _) => {
-                true // FIXME
-            }
-            Format::Map(_f, a) => a.might_match_lookahead(input, next),
-            Format::If(_expr, a, b) => {
-                a.might_match_lookahead(input, next.clone()) || b.might_match_lookahead(input, next)
-            }
-        }
-    }
 }
 
-impl Lookahead {
-    pub fn empty() -> Lookahead {
-        Lookahead { pattern: vec![] }
-    }
-
-    pub fn single(bs: ByteSet) -> Lookahead {
-        Lookahead { pattern: vec![bs] }
-    }
+#[derive(Debug)]
+enum Next<'a> {
+    Empty,
+    Cat(&'a Format, &'a Next<'a>),
+    Tuple(&'a [Format], &'a Next<'a>),
+    Record(&'a [(String, Format)], &'a Next<'a>),
+    Repeat(&'a Format, &'a Next<'a>),
+}
 
-    pub fn alt(a: &Lookahead, b: &Lookahead) -> Lookahead {
-        Lookahead {
-            pattern: Iterator::zip(a.pattern.iter(), b.pattern.iter())
-                .map(|(ba, bb)| ByteSet::union(&ba, bb))
-                .collect(),
+impl Switch {
+    fn reject() -> Switch {
+        Switch {
+            accept: None,
+            branches: vec![],
         }
     }
 
-    pub fn cat(a: &Lookahead, b: &Lookahead) -> Lookahead {
-        let mut pattern = a.pattern.clone();
-        pattern.extend(b.pattern.iter().cloned());
-        Lookahead { pattern }
-    }
-
-    /// Find a lookahead that only the first format matches
-    pub fn new(a: &Format, b: &Format, opt_next: Option<&Format>) -> Option<Lookahead> {
-        const LEN: usize = 2;
-        let next = match opt_next {
-            None => Format::Empty,
-            Some(next) => next.clone(),
-        };
-        let pa = Lookahead::from(a, LEN, next.clone())?;
-        if !b.might_match_lookahead(&pa.pattern, next) {
-            Some(pa)
-        } else {
-            None
+    fn accept(&mut self, index: usize) -> Result<(), ()> {
+        match self.accept {
+            None => {
+                self.accept = Some(index);
+                Ok(())
+            }
+            Some(i) if i == index => Ok(()),
+            Some(_) => Err(()),
         }
     }
-    /*
-        pub fn disjoint(a: &Self, b: &Self) -> bool {
-            for i in 0..std::cmp::min(a.pattern.len(), b.pattern.len()) {
-                if ByteSet::disjoint(&a.pattern[i], &b.pattern[i]) {
-                    return true;
-                }
+
+    fn add_next(&mut self, index: usize, depth: usize, next: &Next) -> Result<(), ()> {
```

Comment:
Is the `depth` the number of bytes of lookahead?

Reply:
Yes, each level of the tree switches on the next byte.